### ISSUE TITLE
Can't configure connection to TFS2017

### ISSUE THREAD

**Issue by Prontsevich** [microsoft/azure-repos-vscode#226]
Hello!

I try to configure your extension and use next settings:
```
{
	"tfvc.location": "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\Common7\\IDE\\TF.exe",
	"team.logging.level": "debug",
	"team.remoteUrl": "http://server:8080/tfs/SuperCollection",
	"team.teamProject": "SuperModule"
}
```
On server I have next structure:
![image](https://cloud.githubusercontent.com/assets/1219707/25966649/473975f6-3694-11e7-9610-68acc366a56b.png)

I opened SuperAngular.WebApp from Dev branch in VSCode and changed any file. But extension doesn't work. What should I do?

I attached [team-extension.log](https://github.com/Microsoft/vsts-vscode/files/994173/team-extension.txt). Also there is no TFVC option in 'Switch SCM Provider...' menu. 
![image](https://cloud.githubusercontent.com/assets/1219707/25967870/373f8092-3698-11e7-9e61-5ae1c4750961.png)
But in the bottom of the window I can see:
![image](https://cloud.githubusercontent.com/assets/1219707/25967971/9541d4ba-3698-11e7-9d06-6ec8043d026a.png)

**Comment by jeffyoung** [comment]
Hi @Prontsevich.  Thanks for trying out the extension!

Can you remove both the `team.remoteUrl` and `team.teamProject` settings? With [those](https://github.com/Microsoft/vsts-vscode#using-external-non-microsoft-source-repositories), you're indicating that your source code is *not* hosted on either Team Services or a Team Foundation Server. Therefore, the extension doesn't even evaluate your `tfvc.location` setting.  (I do see in the logs, however, references to "TFVC" which is inaccurate and confusing, even to me. I'll get those messages fixed.)

I've seen users set both of those settings before (#217) and I'm trying to determine what prompted you to set them? Somewhere I'm leading folks into this trap (of setting these 2 values when they're likely not needed) and I'd like to correct that. I'd definitely appreciate the help!

So please remove those two settings and let me know if you're successful.

**Comment by Prontsevich** [comment]
If I remove them the extension will connect to the incorrect collection. We have several collections on server and extension takes the first one alphabetically.

**Comment by Prontsevich** [comment]
I removed these two settings and put `"tfvc.restrictWorkspace": true`. Extension connected to right project but still doesn't show any changes.

**Comment by jeffyoung** [comment]
So, by default, you were connected to the first team project in your workspace (likely alphabetical). After adding that setting, we use the folder you're running in and map that to a team project and restrict the results to that folder.

Can you run `tf status` from a command prompt in the folder that you're expecting to see changes? That's what the extension so it should show any changes returned.  An example call would be something like this:

`tf status -noprompt -collection:http://java-tfs2015:8081/tfs/DefaultCollection -format:xml -recursive <your-local-path>`

I would expect that command to show the same as what is shown in the viewlet.  If that's true (or better yet, if it isn't), I'd like to know more about your workspace mappings. The results from a `tf workfold` might be helpful.

**Comment by Prontsevich** [comment]
`tf workfold:`
Workspace : PRONTSEVICH (Pronsevich, Sergey)
Collection: http://server:8080/tfs/SuperCollection
 $/CDN: D:\P\CDN
 $/SuperModule: D:\P\SuperModule

**Comment by Prontsevich** [comment]
`tf status`
[brief.txt](https://github.com/Microsoft/vsts-vscode/files/1003674/brief.txt)

**Comment by Prontsevich** [comment]
I changed several files as you can see but in extension I can't see them.

**Comment by jeffyoung** [comment]
Hi Sergey.

Can you delete the existing `team-extension.log` file and retry your scenario with debug logging still enabled? It appears the extension isn't getting initialized for some reason and the last log you had was set up to use External repositories (so the extension never even attempted to initialize the TFVC functionality).

Also, do you have the `tfvc.restrictWorkspace` setting still enabled? If so, what if you directly open the `D:\P\SuperModule` folder (and not, for example, D:\P)?  At that point, I would expect to see some changes since the workspace should be restricted to that team project.  That said, I see in the output of your `tf status` command over 16,000 changes. I suspect it will take some time to parse all of that output and update the viewlet.  So dealing with that many changes at the command line may provide a better experience. However, I will be interested to know what the experience is like with that many changes.

And thanks for providing more information!

**Comment by Prontsevich** [comment]
My workspace settings:
```
{
	"vsicons.presets.angular": true,
	"files.exclude": {
		"bin/": true,
		"dist/": true,
		"obj/": true,
		"Properties/": true,
		"*.csproj*": true,
		"*.sln": true
	},
	"tfvc.location": "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\Common7\\IDE\\TF.exe",
	"tfvc.restrictWorkspace": true,
	"team.logging.level": "debug"
}
```
I signed in using my username and password and got next log file [team-extension.txt](https://github.com/Microsoft/vsts-vscode/files/1009966/team-extension.txt)

**Comment by Prontsevich** [comment]
As I can see in log I have error with authorization. But I entered correct credentials.

**Comment by jeffyoung** [comment]
Thanks Sergey.

I had a look in your log file and my guess is that you're using a server workspace (since the log shows a TF30063 error (`TFVC errors: TF30063: You are not authorized to access server\SuperCollection`).  Could you send a screenshot of your workspace by following these [instructions](https://github.com/Microsoft/vsts-vscode/blob/master/TFVC_README.md#what-is-the-difference-between-a-local-and-server-workspace-how-can-i-tell-which-one-im-working-with)? I just need to know the value of the 'Location' property.

What I'm investigating now is if I can determine why the `tf workfold` command we run doesn't also throw this error in your scenario (assuming you actually do have a Server workspace). In all instances up to now, we've seen it after workfold.  However, your workfold runs just fine but `tf resolve` fails. When I force an exception during resolve here locally, I get an unhandled rejected promise (which I'll need to fix) and don't see any error message.  Now when I detect this condition with workfold, I show this error message:

![image](https://cloud.githubusercontent.com/assets/2796865/26207820/999463de-3bb6-11e7-939e-9b32324e1221.png)

If you do have a Server workspace, you should be able to get working if you can switch over to a Local workspace (here's some [doc](https://www.visualstudio.com/en-us/docs/tfvc/decide-between-using-local-server-workspace) on the differences).  Granted, not everyone can do that but it's an option. Otherwise, I'm tracking this issue in #176.

**Comment by jeffyoung** [comment]
As of v1.118.1 (just released), this scenario should be detected and an alert displayed (as in my previous response).  I'd still like to know more about what's going on however (if you're still interested).

**Comment by Prontsevich** [comment]
Hello :)
Here my screenshot of `tf workspace`:  ![workspace](https://cloud.githubusercontent.com/assets/1219707/26348580/cda796a2-3fb5-11e7-9dad-93be15a8b3ef.PNG)

I updated to the latest version of extension and got the message about server workspace.

**Comment by jeffyoung** [comment]
OK.  😕 

Can we see what the same command that's failing in the extension does at the command line?  The command to run would be `tf resolve -noprompt d:\P\SuperModule\SuperAngular\Dev\SuperAngular.WebApp -recursive -preview`

**Comment by Prontsevich** [comment]
This is the result: `There are no conflicts to resolve.`

**Comment by jeffyoung** [comment]
So the difference here seems to be the passing of the `-login` switch on the command line.  The extension passed it in the form `-login:username,password` and I had you run it at the command line without providing them.

Could you now run the following (notice the adding of the `-login` switch):
`tf resolve -noprompt -login:<username>,<password> d:\P\SuperModule\SuperAngular\Dev\SuperAngular.WebApp -recursive -preview`

Obviously provide your username and password (that you did when you first signed in).  I'm wondering if there's something in the way we're passing the username and password.

And continued thanks for hanging in there with me!

**Comment by Prontsevich** [comment]
Still the same result: `There are no conflicts to resolve.`

**Comment by Prontsevich** [comment]
I have username like 'DOMAIN\sergey.prontsevich'.

**Comment by Prontsevich** [comment]
Any news?

**Comment by jeffyoung** [comment]
Hi @Prontsevich.

I've been doing a bit more investigation and looking for ways to see if I can provide guidance for more debugging on your end.  Unfortunately, I've got nothing to report yet.  However, can you tell me if there's anything special/different about your password?  At this point, I’m trying to figure out if something is going on when we pass credentials into the tf.exe process via stdin that could be causing it.

I have another report of this issue (exactly this issue) and, so far, the only difference appears to be that the machine being used (which is running VSCode) is on one domain and the TFS server is on another.  But the fact we're running the exact same command line either via the command prompt or within VS Code and getting different results leads me to believe that it's something in how we're passing credentials.

Can you also confirm (again) that you're able to get access to the build and work item information in the status bar?  When trying to reproduce this on my own, if I supplied an invalid password, I didn't even get to the point of running `tf resolve` (I couldn't get the build or work item information).  Otherwise, I couldn't reproduce it.

From an investigation standpoint, would you be willing to try and use the TEE CLC (which requires Java) to see if the behavior is the same?  If so, the instructions are [here](https://github.com/Microsoft/vsts-vscode/blob/master/TFVC_README.md#how-do-i-set-up-the-clc-on-windows).  Otherwise, I'd like to put together a private build of the extension that will dump out the credentials used into the log file.  I'd just like to know that what we're passing into the tf process is what we expect.  LMK if you'd be willing to try that as well.

Thanks @Prontsevich.

**Comment by Prontsevich** [comment]
Hello!
When I started VS Code I can see next info in status bar 
![capture](https://cloud.githubusercontent.com/assets/1219707/26732497/756c3f84-47c0-11e7-8bd6-09e5b1b71bc6.png)

An my machine has another domain then on TFS server.

**Comment by jeffyoung** [comment]
Thanks for the information @Prontsevich.  Looks like I'll need to try and set up a local repro (with one machine on one domain and the TFS server on another).

**Comment by jeffyoung** [comment]
So I set up a TFS 2017 server on a domain that my dev machine is not a part of.  But I did put it on the same domain that my user account is a part of.  Unfortunately, I was unable to repro the issue. 😖 

I then poked around the code behind tf.exe as well as the server code for QueryConflicts.  There, it appeared that an HTTP Status of 302 might cause the issue (but I'm not convinced).

I then noticed that both you and Pat (the other customer with this issue) had only "server/collection" in the Server portion of the workspace dialog.  Since this is what was being reported in the exception message, I thought it might have something to do with it.  I had set up the same scenario (or so I thought) but could always get the first `tf resolve` to return 0 instead of 100.  I have noticed that the TF.exe from VS2017 will show the full URL for the workspaces while the TF.exe from VS2015 (which both you and Pat are using) will only show "server/collection".  However, again, that doesn't seem to be the cause.

I would still be interested to know if using the TEE CLC ends up with the same exception.  In addition, it would be interesting to know if TF.exe from VS2017 ended up with the same exception.  [Installing TF.exe from VS 2017](https://github.com/Microsoft/vsts-vscode/blob/master/TFVC_README.md#how-can-i-acquire-tfexe-do-i-need-a-version-of-visual-studio) would be easier than the [TEE CLC](https://github.com/Microsoft/vsts-vscode/blob/master/TFVC_README.md#how-do-i-set-up-the-clc-on-windows).  And if either of them is tried, you'd have to run the ["workspaces -collection:url"](https://github.com/Microsoft/vsts-vscode/blob/master/TFVC_README.md#using-the-tee-clc-i-am-unable-to-access-an-existing-local-workspace-what-can-i-do) option in order for that version of the TF command line to discover the workspace.

**Comment by Prontsevich** [comment]
I installed TF from VS2017 and it didn't help me. Then I tried  TEE CLC and got **"No Team Services or Team Foundation Server repository configuration was found. Ensure you've opened a folder that contains a repository."**

**Comment by jeffyoung** [comment]
Hi Sergey.  Thanks for getting back to me.  I'm losing my optimism that we'll be able to get the extension working with your network setup.  I just don't know what the problem is (especially concerning the fact that the command line is the same between what's run at a command prompt (which works) and the extension (which doesn't)).

However, to get the CLC to detect the workspace created by Visual Studio (or TF.exe) you need to follow [these instructions](https://github.com/Microsoft/vsts-vscode/blob/master/TFVC_README.md#using-the-tee-clc-i-am-unable-to-access-an-existing-local-workspace-what-can-i-do). Once you run `tf workspace` with the CLC the local cache that CLC uses will have information about the workspace and it will then be able to (try and) connect.  I don' t have a lot of hope here either but let's see what happens.

**Comment by Prontsevich** [comment]
I was able to use CLC :) but when open my SuperModule project folder extension connected to CDN project.
`tf workfold` returns:
```
Workspace:  PRONTSEVICH
Collection: http://server:8080/tfs/SuperCollection/
 $/CDN: D:\P\CDN
 $/SuperModule: D:\P\SuperModule
 $/ArrangementServer/Dev: D:\P\ArrangementServer\Dev
 $/ArrangementServer/Documents: D:\P\ArrangementServer\Documents
```
**"tfvc.restrictWorkspace": true**

**Comment by jeffyoung** [comment]
Thanks @Prontsevich.  I'm glad that the CLC is working for you (although it's quite a bit concerning to me that it does). 😕 

That said, I had a look at restrictWorkspace and I see the problem (let me see if I can come up with a fix). It turns out that no matter what folder you pass to the `workfold` command via the CLC, the CLC will always return *all* of the mappings (like you're showing in the last posting) instead of the mapping that applies to the folder that was passed in.  tf.exe always returns just the mapping that applies to the folder you provide at the command line.  In this case, that would be the second line (and, thus, the extension would pick the proper team project).

I'd be curious what would happen if you opened `D:\P\SuperModule` directly (and not any sub-folder).  While that's not what you want, I expect that it would pick the proper team project.

**Comment by Prontsevich** [comment]
I installed this private build. And finally extension connected to the correct project :) But I still can't see any changes in Source Control tab. There is no TFVC provider option to select, only Git. So, now extension connected but still doesn't work :(

**Comment by jeffyoung** [comment]
OK, thanks for the update Sergey.  Can you provide a new log that I can take a look at?  Now that you're using the CLC, I wonder if you might be running into [this issue](https://github.com/Microsoft/vsts-vscode/issues/260)?  If the project is large, then I would suspect it could be taking a while.

**Comment by Prontsevich** [comment]
You're right :) But now I have very long list of excluded items (mainly node_modules files). How can I hide them?

**Comment by jeffyoung** [comment]
Do you have a .tfignore file?  Presumably not but one should work in this case.  For some details, I made some comments on another thread [here](https://github.com/Microsoft/vsts-vscode/issues/248#issuecomment-307469405).

**Comment by Prontsevich** [comment]
No, I don't have it. And I don't see any option in menu for creating it as in docs.

**Comment by Prontsevich** [comment]
When will be the new version for extension? I don't know why, but VS code reinstalls extension v119 after I installed v120 :(

**Comment by jeffyoung** [comment]
You can create the .tfignore manually by copying the example at https://www.visualstudio.com/en-us/docs/tfvc/add-files-server#tfignore-file-example and dropping it into the root of your repository. (Those docs are for the Visual Studio IDE, not Visual Studio Code).  Then, you should add `\node_modules` to the .tfignore. Visual Studio will show all of your files as 'Detected Changes' but doesn't show them individually in the UI.  Adding a .tfignore file will reduce the number of 'Detected Changes' shown in Visual Studio and will also reduce the number of files you see in the TFVC SCM Viewlet in VS Code.  (The .tfignore file provides basically the same functionality as .gitignore does for Git.)  When using an ignore file, you're basically telling the SCM provider what files you aren't interested in tracking via source control.

Also, v119 is being reinstalled because I use 0 as the major number for private drops (as a quick indication that it is a private drop).  Since **1**.119.0 is 'greater' than **0**.120.0 (according to semver), 1.119.0 is the latest version.  I'll be releasing another update some time today (1.119.1).

Let me know how it goes after adding a .tfignore file (oh, and you should check it in too).

**Comment by Prontsevich** [comment]
I created file manually and everything works now. Thank you :)
But now I have a complaint about the speed of work. It takes toooo long time to init extension when I opening my project and project isn't big. And I don't see current state of change files in Source control tab after opening. I need to change any file to recalculate states and then will see all info. Also extension needs 5-10 sec to refresh Source control tab after saving changes in one file. 
Is it possible to optimize extension that it works line in Visual studio?

**Comment by jeffyoung** [comment]
Hi Sergey. I'm glad you got the .tfignore file working.  I think the Visual Studio IDE just "works around" this issue by hiding them under "Detected Changes". At this time, the VS Code extension doesn't have a similar concept (I don't think it's possible to create a group in the SCM viewlet showing a number of files without actually showing the files as well, although I'd need to investigate).

Unfortunately, the Java-based TEE CLC isn't super-fast (each time it is called a new Java VM has to be created). With both the CLC and TF.exe, we always create an extra process to keep around for the next call we need to make to the server.  So when the extension is first run, and we detect TFVC, we spin up either TF.exe or the CLC and have it waiting until we need to make a call. Once we make a call, we start another to give it time to initialized while the first process is doing its work.  I'll investigate what you're seeing with regards to "I don't see current state of change files in Source control tab after opening. I need to change any file to recalculate states and then will see all info." That certainly sounds like a bug.

Finally, it shouldn't take 5-10 seconds to refresh although a comparison between what the extension does and the same command line would be telling. Would it be possible for you to try the same command at the command line and provide the results? The extension logs the time it takes for each call in the team-extension.log file so we can get that data from there. As far as making it as fast as the VS IDE, I don't believe the VS IDE uses the TF.exe command line (it certainly doesn't use TEE) so it wouldn't suffer from spinning up processes for each call to the server (and may very well cache a lot of information in the background).  TFVC support in the VS IDE has been around for quite a few years so it has a bit of a head start on the Code extension. Like I said, 5-10 seconds per operation shouldn't be the expectation but I'd certainly like to know how the extension compares with the command line.

One crude way to determine how long it takes at the command line would be to do something like this: `echo %time% & <command line> & echo %time%`

**Comment by jeffyoung** [comment]
Hi Sergey.  Based on your comment way back in this thread at https://github.com/Microsoft/vsts-vscode/issues/226#issuecomment-305820241, you're in the siutation where your client machine is on one domain and the TFS server is on another.  With TF.exe, there was a bug in which we weren't passing the domain within the credentials sent to TF.exe.  A fix for that is now in v1.122.0 which I just released.  In short, you should now be able to go back to using TF.exe (and I'd be interested in knowing if there's a performance improvement between the CLC and TF.exe).
